File: basics/biconnected_components.py
import sys
N, M = [int(x) for x in sys.stdin.readline().split()]
adj = [[] for x in range(N+1)]
for i in range(M):
  a, b = [int(x) for x in sys.stdin.readline().split()]
  adj[a].append(b)
  adj[b].append(a)


visited = [False for x in range(N+1)]
disc = [0 for x in range(N+1)]
low = [float("inf") for x in range(N+1)]
time = 0
components = []


def dfs(curr, prev):
  global time 
  visited[curr] = True
  time += 1
  disc[curr] = time
  low[curr] = time
  children = 0

  for node in adj[curr]:
    if node != prev:
      if visited[node] and low[curr] > disc[node]:
        bi_comp.append(list([curr, node]))
        low[curr] = min(low[curr], disc[node])
      elif not visited[node]:
        bi_comp.append(list([curr, node]))
        children += 1
        dfs(node, curr)
        low[curr] = min(low[curr], low[node])
        if prev == -1 and children > 1:
          temp = []
          while len(bi_comp) > 0:
            if bi_comp[-1] == [curr, node]:
              temp.append(bi_comp.pop(-1))
              break
            temp.append(bi_comp.pop(-1))
          components.append(temp)
        elif low[node] >= disc[curr]:
          temp = []
          while len(bi_comp) > 0:
            if bi_comp[-1] == [curr, node]:
              temp.append(bi_comp.pop(-1))
              break
            temp.append(bi_comp.pop(-1))
          components.append(temp)



for i in range(0, N):
  if not visited[i]:
    bi_comp = []
    dfs(i, -1)

for i in components:
  print(i)

File: basics/biconnected_components.cpp
#include <bits/stdc++.h>
using namespace std;
const int inf = 0x3f3f3f3f;
int N, M, a, b, t = 0, low[200000], disc[200000];
bool vis[200000];
vector<int> adj[200000];
vector<vector<int>> s;
vector<vector<vector<int>>> components;


void dfs(int curr, int prev){
  vis[curr] = true;
  t++;
  disc[curr] = t;
  low[curr] = t;
  int children = 0;
  for (auto node: adj[curr]){
    if (prev != node){
      if (vis[node] && low[curr] > disc[node]){
        low[curr] = min(low[curr], disc[node]);
        s.push_back({curr, node});
      } else if (!vis[node]){
        s.push_back({curr, node});
        children++;
        dfs(node, curr);
        low[curr] = min(low[curr], low[node]);
        if (prev == -1 && children > 1){
          vector<vector<int>> temp;
          while (s.size() > 0){
            if (s.back() == vector<int>{curr, node}){
              temp.push_back(s.back()); s.pop_back();
              break;
            }
            temp.push_back(s.back()); s.pop_back();
          }
          components.push_back(temp);
        } else if (low[node] >= disc[curr]){
          vector<vector<int>> temp;
          while (s.size() > 0){
            if (s.back() == vector<int>{curr, node}){
              temp.push_back(s.back()); s.pop_back();
              break;
            }
            temp.push_back(s.back()); s.pop_back();
          }
          components.push_back(temp);
        }
      }
    }
  }
}


int main(){
  memset(low, inf, sizeof(low));
  memset(disc, 0, sizeof(disc));
  memset(vis, false, sizeof(vis));
  cin >> N >> M;
  for (int i = 0; i < M; i++){
    cin >> a >> b;
    adj[a].push_back(b);
    adj[b].push_back(a);
  }

  for (int i = 0; i < N; i++){
    if (!vis[i]){
      s.clear();
      dfs(i, -1);
    }
  }

  for (auto component: components){
    for (auto edge: component){
      for (auto node: edge){
        cout << node << " ";
      }cout << " | ";
    } cout << " \n";
  }

  return 0;
}